Productionise

SiAdcock opened the pull request:
- Uses `micro` and `pm2` to start a production web server on port 9000 (standard for Guardian services)
- Code splits in `server.js` to dynamically load bundle for current page

MatthewJWalls commented:
I get this when I try to run this branch locally, any idea?

![image](https://user-images.githubusercontent.com/1218561/36552511-f5bbf4ac-17f1-11e8-8285-0c6aa2ccdf7d.png)

SiAdcock commented:
@MatthewJWalls Try running

```
$ yarn build && yarn start
```

sndrs commented on src/server.js:
```diff
@@ -3,14 +3,13 @@
-
 import doc from 'lib/__html';
 
-const pages = requireDir('./pages');
```
maybe comment this? or do you think we'll need to swap it out for good?

SiAdcock replied:
I'm working on an alternative using a dynamic `import()`. Give me a moment and I'll push my changes, just need to make sure I haven't broken the dev server!

sndrs commented:
should we add something to the readme?

sndrs commented on cloudformation.yml:
```diff
@@ -0,0 +1,208 @@
+
+          cd /home/ubuntu
+
+          curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
```
is there any way we can use nvm here instead?

SiAdcock replied:
Good call. I'm probably gonna spend ages tweaking this file TBH.

sndrs replied:
yeah travis uses nvmrc – even if we just use to get eh version to install rather than full nvm (which may be overkill?) will save someone in the future having to work out why they havn't updated node on the server :)

wonder if we could dockerise?

sndrs replied:
dockerfy

sndrs commented on cloudformation.yml:
```diff
@@ -0,0 +1,208 @@
+          git clone https://github.com/guardian/guui.git
+          cd guui
+
+          yarn
```
need to install yarn?

SiAdcock replied:
Oh god yes

MatthewJWalls replied:
I wouldn't worry too much about getting the CF file perfect under this PR, it'll probably take a while of tweaking until it installs properly on an ec2 instance anyway (making sure it finds nvm/node/yarn properly etc). And it's missing the scaling groups that I need to add.

sndrs commented:
does `cloudformation.yml` need to be in top level?

SiAdcock commented:
@sndrs Thanks for reviewing. I'll find a nice spot for `cloudformation.yml` to keep the the repo tidier. Maybe a top level `cfn` directory? Or I could create a op level `__config__` directory for stuff like Webpack config, `cloudformation.yml` and god knows what else.

sndrs commented:
@SiAdcock `__config__` ftw

SiAdcock commented:
`#underscores4life`

SiAdcock commented on src/server.js:
```diff
@@ -1,16 +1,19 @@
-export default (state: { page: string }): string => {
-    const Page = pages[state.page].default;
+const renderPage = async function renderPage(pageName) {
+    const page = await import(`./pages/${pageName}`);
```
@sndrs What do you think of code splitting here instead of requiring the `pages` directory?

sndrs replied:
does `import` work like this in node?

sndrs replied:
i went with the whole dir import because it would put all the modules into memory once this file was `required`. maybe that was a bit preemptive?

SiAdcock replied:
Okay, that's a good point! I'll change it back

RichieAHB replied:
If you ever want to do the importing client side then you’ll probably have to keep this as a dynamic import and use the Airbnb dynamic import plugins for webpack and node in the different environments respectively. If you’re only ever server rendering then using Node specific importing will be fine.

SiAdcock commented on src/server.js:
```diff
@@ -23,3 +26,15 @@ export default (state: { page: string }): string => {
+            path.join('dist', req.url.split('/assets/javascript/')[1]),
+            'utf8',
+        );
+    }
```
A bit mad we have to implement our own `serve-static` middleware! Maybe `micro` is a bit too simple for this use case?

sndrs replied:
would assets not be served from the CDN as normal?

SiAdcock replied:
That's a good point too! I guess we just need to do this in our dev server

SiAdcock replied:
Yeah, true. I guess this is fine as a first step, next we can look at how to deploy, where to host static assets.

SiAdcock commented:
Hi @sndrs, i believe I've addressed most of your points.

I have added a `pagesConfig` to `webpack.config.js` which bundles pages for the server (they are already included in the browser bundle. `pagesConfig` is used for preserving your `requireDir` functionality, by creating a `pages` directory under `dist`. Let me know if you have any further thoughts!

sndrs commented on webpack.config.js:
```diff
@@ -35,11 +35,20 @@ const envConfig = production
-const platformConfig = platform =>
-    merge.smart(
+const platformConfig = platform => {
+    const lib =
```
think this could go in `serverConfig`?

SiAdcock replied:
Yeah maybe, I’ll take a look

sndrs commented on webpack.config.js:
```diff
@@ -59,9 +69,27 @@ const serverConfig = merge.smart(
     },
     platformConfig('server'),
 );
+const pagesConfig = merge.smart(
```
not too sure what's happening here?

SiAdcock replied:
This will generate a bundle for each page. The server requires each bundle under the dist/pages directory using requireDir

sndrs replied:
it looks like i tried that at some point! https://github.com/guardian/guui/blob/master/.babelrc.js#L43 

i sounds like a good idea to me?

SiAdcock commented:
This is becoming difficult to grok. I'm going to break it down
